tutorials/sphinx-tutorials/multiagent_competitive_ddpg.py

```diff
@@ -0,0 +1,992 @@
+        ],  # Write ``(group, "state_action_value")``
+    )
+
+    critics[group] = TensorDictSequential(
+        cat_module, critic_module
+    )  # Run them in sequence
+
+
+######################################################################
+# Let us try our policy and critic modules. As pointed earlier, the usage of
+# :class:`TensorDictModule` makes it possible to directly read the output
+# of the environment to run these modules, as they know what information to read
+# and where to write it.
+#
+# We can see that after each group's networks are run their output keys are added to the data under the
+# group entry.
+#
+# **From this point on, the multi-agent-specific components have been instantiated, and we will simply use the same
+# components as in single-agent learning. Isn't this fantastic?**
+#
+reset_td = env.reset()
+for group, _agents in env.group_map.items():
+    print(
+        f"Running value and policy for group '{group}':",
+        critics[group](policies[group](reset_td)),
+    )
+
+######################################################################
+# Data collector
+# --------------
+#
+# TorchRL provides a set of data collector classes. Briefly, these
+# classes execute three operations: reset an environment, compute an action
+# using the policy and the latest observation, execute a step in the environment, and repeat
+# the last two steps until the environment signals a stop (or reaches a done
+# state).
+#
+# We will use the simplest possible data collector, which has the same output as an environment rollout,
+# with the only difference that it will auto reset done states until the desired frames are collected.
+#
+# We need to feed it our exploration policies. Furthermore, to run the policies from all groups as if they were one,
+# we put them in a sequence. They will not interfere with each other as each group writes and reads keys in different places.
+#
+
+# Put exploration policies from each group in a sequence
+agents_exploration_policy = TensorDictSequential(*exploration_policies.values())
+
+collector = SyncDataCollector(
+    env,
+    agents_exploration_policy,
+    device=device,
+    frames_per_batch=frames_per_batch,
+    total_frames=total_frames,
+)
+
+######################################################################
+# Replay buffer
+# -------------
+#
+# Replay buffers are a common building piece of off-policy RL sota-implementations.
+# There are many types of buffers, in this tutorial we use a basic buffer to store and sample tensordict
+# data randomly.
+#
+
+replay_buffers = {}
+for group, _agents in env.group_map.items():
+    replay_buffer = ReplayBuffer(
+        storage=LazyTensorStorage(
+            memory_size, device=device
+        ),  # We will store up to memory_size multi-agent transitions
+        sampler=RandomSampler(),
+        batch_size=train_batch_size,  # We will sample batches of this size
+    )
+    replay_buffers[group] = replay_buffer
```

Comment:
what is preventing us from building a single RB?

Reply:
we do not want to mix data from different groups. We may decide to collect more for one group or smth like that